Feature/single topic multi tenant (#802)
permitio/opal: documentation/docs/tutorials/single-topic-multi-tenant.mdx

---
sidebar_position: 12
title: Single-Topic Multi-Tenant Configuration
---

# Single-Topic Multi-Tenant Configuration

This document describes how to configure OPAL for multi-tenant environments using a single topic approach. This configuration allows adding new tenants without restarting OPAL services.

> ## Understanding Single-Topic Multi-Tenancy

Author: Yes, great - that's gonna be far more accurate.

```

This traditional approach creates several challenges:

- **Event-driven vs polling limitations** - When you need event-driven data updates instead of interval-based polling, managing multiple topics becomes complex and resource-intensive.

- **Bundle scalability issues** - For thousands of tenants with large datasets, updating data through undivided bundles becomes impractical. This forces you to implement data sharding or tenant-specific bundle divisions, which are difficult to prepare and maintain.

- **State management complexity** - Divided bundles introduce problems with state management, particularly around bundle ordering and ensuring consistent data application across tenants.

- **Topic proliferation** - Achieving tenant isolation through separate data sources results in an ever-growing number of topics, making the system difficult to manage and scale.

- **Configuration management overhead** - Each new tenant requires configuration changes and service restarts, making dynamic tenant onboarding cumbersome.

The single-topic configuration described below addresses these issues by providing a simpler, more scalable approach.

> 

Reviewer: Worth adding a diagram to explain the architecture

```suggestion

```mermid
flowchart
    subgraph "Multi-scope"
        DS1["Data source: Tenant 1"]
        DS2["Data source: Tenant 2"]
        DSN["Data source: Tenant N"]
        DS1 --> OC1["OPAL Client: Scope 1"]
        DS2 --> OC2["OPAL Client: Scope 2"]
        DSN --> OCN["OPAL Client: Scope N"]
        OC1 <--> SVC["Application services"]
        OC2 <--> SVC
        OCN <--> SVC
    end
```

```

### Single-topic configuration

With single-topic configuration, all tenants share one topic:

```
All tenants: OPAL_DATA_TOPICS=tenant_data  # no restart needed when adding tenants
```

The key insight is that instead of creating separate topics for each tenant, this approach dynamically registers **multiple external data sources** under a single topic. Each tenant gets its own data source configuration, but all sources publish to the same `tenant_data` topic:

- **Tenant 1**: External data source A → `tenant_data` topic → `/acl/tenant1` path
- **Tenant 2**: External data source B → `tenant_data` topic → `/acl/tenant2` path  
- **Tenant N**: External data source N → `tenant_data` topic → `/acl/tenantN` path

> This allows for:
> - **Dynamic data source registration** via `/data/config` API calls
> - **Tenant isolation** through unique destination paths in OPA
> - **Unified topic management** while maintaining separate data sources
> - **Real-time tenant onboarding** without service restarts
> 
> ## Benefits
> 
> - **No service restarts required when adding new tenants** - Unlike configurations with static data sources where `OPAL_DATA_CONFIG_SOURCES` is fixed at startup, this approach uses dynamic data source management via the `/data/config` API endpoint.
> 
> - **Better resource utilization** - Compared to PATCH-based approaches that require maintaining incremental state changes, this configuration always recreates the complete data state for each tenant, eliminating the complexity of managing partial updates and potential inconsistencies.
> 
> - **Simplified topic management** - Single topic eliminates the need to manage an ever-growing list of tenant-specific topics.
> 
> - **Real-time data source notifications** - Changes to tenant data can be immediately propagated using `/data/config` API calls, which notify all subscribed clients without requiring bundle regeneration or service restarts.
> 
> - **Complete state reconstruction** - Unlike bundle-based methods that may require complex state merging, each data update via `/data/config` ensures full data consistency by reconstructing the complete tenant state.
> 
> - **Faster tenant onboarding** - New tenants can be added instantly through API calls rather than configuration file changes and service restarts.

Reviewer: ```suggestion
### Benefits

- **No service restarts required when adding new tenants** - does not require `OPAL_DATA_CONFIG_SOURCES` configuration changes, with a rolling restart.
- **Better resource utilization** - tenants are accessible from all OPAL Clients (i.e. OPA instances).
- **Simplified topic management** - no need to manage new topics or scopes being in sync with tenants.
- **Real-time data source notifications** - Changes to tenant data can be immediately propagated using triggered data updates, propagating to all OPAL Clients without requiring restarts.
- **Faster tenant onboarding** - New tenants can be added instantly without configuration changes or new deployments.

### Challenges

- **Higher memory footprint per client** - every OPAL Client must hold *all* tenants' data, which can reduce performance from OPA, or exceed it limits for larger tenants.
- **Increased Data Source load** - all OPAL Clients would need to fetch the tenants' data, even when it is used by only some of them.
- **Horizontal-scaling Data** - supporting tenants with larger datasets may exceed OPA's data capacity.
- **Horizontal-scaling Tenants** - supporting more tenants would require larger OPAL Client instances.
```

> 

Reviewer: ```suggestion

```mermid
flowchart
    subgraph "Single-topic"
        DS1["Data source: Tenant 1"]
        DS2["Data source: Tenant 2"]
        DSN["Data source: Tenant N"]
        DS1 --> OCA["OPAL Client: All tenants"]
        DS2 --> OCA
        DSN --> OCA
        OCA <--> SVC["Application services"]
    end
```

```

> ## Prerequisites
> 
> - Docker and Docker Compose
> - `jq` for JSON processing
> 
> ## Configuration
> 
> ### OPAL Client Configuration

Reviewer: Incorrect title hierarchy
```suggestion
## Example Setup - Docker Compose

Provided a docker compose to show an example setup of a single-topic multi-tenant architecture. 

### Prerequisites

- Docker and Docker Compose
- `jq` for JSON processing

### Configurations
#### OPAL Client Configuration
```

> 

Reviewer: Consider also adding a mermid chart explaining the setup

```mermid
flowchart TD
 subgraph OPAL["OPAL control plane"]
        server["OPAL Server"]
        client["OPAL Client"]
        opa["OPA (policy engine)"]
  end
    GitHub["Policy Repo (GitHub)"] --> server
    server -- broadcast updates --> broadcast["Postgres"]
    server -- trigger data update /acl/{tenant} --> client
    client -- fetch /acl/{tenant} data --> provider["NGINX external API"]
    client -- store data at /acl/{tenant} --> opa
    opa <-- authz query from /acl/{tenant} --> App{"Your services / API"}
```

```yaml
opal_client:
  environment:
    - OPAL_DATA_TOPICS=tenant_data  # single topic for all tenants
    - OPAL_DATA_UPDATER_ENABLED=true
```

> ### OPAL Server Configuration

Reviewer: ```suggestion
#### OPAL Server Configuration
```

```yaml
opal_server:
  environment:
    # Initially empty - tenants added via API
    - 'OPAL_DATA_CONFIG_SOURCES={"config": {"entries": []}}'
    
    # Policy repository configuration
    - OPAL_POLICY_REPO_URL=https://github.com/plduser/opa-zero-poll-single-topic-multi-tenant.git
    - OPAL_POLICY_REPO_PATH=docker/docker_files/policies
```

> ## How it works
> 
> ### Data source entries

Reviewer: ```suggestion
### How it works

#### 1. Data source entries
```

Each tenant's data is configured as a separate data source entry with:
- **url**: Endpoint serving tenant data
- **topics**: Always `["tenant_data"]` for all tenants
- **dst_path**: Unique path for tenant data isolation (e.g., `/acl/tenant1`)

> ### Topic publishing

Reviewer: ```suggestion
#### 2. Topic publishing
```

When any tenant data is updated:
1. OPAL server publishes to the single `tenant_data` topic
2. All OPAL clients subscribed to `tenant_data` receive the notification
3. Clients fetch and update only the specific tenant data that changed

> ### Data isolation

Reviewer: ```suggestion
#### 3. Tenant Data isolation
```

```

OPAL Client logs:
```bash
docker compose -f docker-compose-single-topic-multi-tenant.yml logs -f opal_client
```

Look for these log patterns:

**Server side:**
```
Publishing data update to topics: {'tenant_data'}
Broadcasting incoming event: {'topic': 'tenant_data'}
```

**Client side:**
```
Received notification of event: tenant_data
Fetching data from url: http://example_external_data_provider:80/acl/tenant1
Saving fetched data to policy-store: destination path='/acl/tenant1'
```

## Data Updates

### Update existing tenant data

```bash
curl -X POST http://localhost:7002/data/config \
  -H "Content-Type: application/json" \
  -d '{
    "entries": [{
      "url": "http://example_external_data_provider:80/acl/tenant1",
      "topics": ["tenant_data"],
      "dst_path": "/acl/tenant1"
    }],
    "reason": "Refresh tenant1 data"
  }'
```

## Testing Authorization

### Test tenant-specific policies

```bash
curl -X POST http://localhost:8181/v1/data/policies/rbac/allow \
  -H "Content-Type: application/json" \
  -d '{
    "input": {
      "user": "alice",
      "action": "read",
      "resource": "document1",
      "tenant_id": "tenant1"
    }
  }' | jq .
```

> ## Production Considerations

Reviewer: This is slight duplicate of other articles on the OPAL docs, but its good to place it here too

> ### Use webhooks instead of polling

Reviewer: ```suggestion
### Use Policy Repo webhooks instead of polling
```

```

### Secure data sources

Configure authentication for external data APIs:

```bash
curl -X POST http://localhost:7002/data/config \
  -H "Content-Type: application/json" \
  -d '{
    "entries": [{
      "url": "https://secure-api.yourdomain.com/tenant-data/tenant1",
      "config": {
        "headers": {
          "Authorization": "Bearer your-secure-token"
        }
      },
      "topics": ["tenant_data"],
      "dst_path": "/acl/tenant1"
    }]
  }'
```

### Scale horizontally

Increase worker count for higher load:

```yaml
opal_server:
  environment:
    - UVICORN_NUM_WORKERS=4
```

## Troubleshooting

### Tenant not receiving data

Check:
1. Topic configuration: `OPAL_DATA_TOPICS=tenant_data`
2. External data provider endpoint availability
3. OPAL Client logs for connection errors

```bash
# Test data endpoint
curl http://localhost:8090/acl/tenant1

# Check client logs
docker logs $(docker compose ps -q opal_client)
```

### Data isolation issues

Verify:
1. Unique `dst_path` for each tenant
2. Policies correctly handle `tenant_id` in input

```bash
# Check OPA data structure
curl -s http://localhost:8181/v1/data | jq .
```

> ## Comparison with Traditional Approach
> 
> | Aspect | Traditional | Single-Topic |
> |--------|-------------|--------------|
> | Topics per 1000 tenants | 1000 | 1 |
> | Service restart required | Yes | No |
> | Memory usage growth | O(N²) | O(N) |
> | Configuration complexity | High | Low |

Reviewer: This is a duplicate of the info stated above, and contain some vague terms. I would omit that from the tutorial as it is not needed.

```suggestion
```

      "topics": ["tenant_data"],
      "dst_path": "/acl/tenant1"
    }],
    "reason": "Refresh tenant1 data"
  }'
```

## Testing Authorization

### Test tenant-specific policies

```bash
curl -X POST http://localhost:8181/v1/data/policies/rbac/allow \
  -H "Content-Type: application/json" \
  -d '{
    "input": {
      "user": "alice",
      "action": "read",
      "resource": "document1",
      "tenant_id": "tenant1"
    }
  }' | jq .
```

## Production Considerations

### Use Policy Repo webhooks instead of polling

For production environments, configure webhooks for immediate policy updates:

```yaml
opal_server:
  environment:
    - OPAL_POLICY_REPO_WEBHOOK_SECRET=your-secure-secret
```

### Secure data sources

Configure authentication for external data APIs:

```bash
curl -X POST http://localhost:7002/data/config \
  -H "Content-Type: application/json" \
  -d '{
    "entries": [{
      "url": "https://secure-api.yourdomain.com/tenant-data/tenant1",
      "config": {
        "headers": {
          "Authorization": "Bearer your-secure-token"
        }
      },
      "topics": ["tenant_data"],
      "dst_path": "/acl/tenant1"
    }]
  }'
```

> ### Scale horizontally

Reviewer: This is not Horizontal Scaling, it is inside the same pod / server 😬
Adding additional Uvicorn workers can help utilizing the server resources better, useful for Vertical Scaling when increasing the CPU / Memory of the server.
